server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -0,0 +1,245 @@
+            IOUtils.closeWhileHandlingException(current);
+            IOUtils.closeWhileHandlingException(readers);
+            throw e;
+        }
+    }
+
+    @Override
+    public boolean ensureSynced(Location location) throws IOException {
+        try (ReleasableLock ignored = writeLock.acquire()) {
+            assert location.generation <= current.getGeneration();
+            if (location.generation == current.getGeneration()) {
+                ensureOpen();
+                return prepareAndUpload(primaryTermSupplier.getAsLong(), location.generation);
+            }
+        } catch (final Exception ex) {
+            closeOnTragicEvent(ex);
+            throw ex;
+        }
+        return false;
+    }
+
+    @Override
+    public void rollGeneration() throws IOException {
+        syncBeforeRollGeneration();
+        if (current.totalOperations() == 0 && primaryTermSupplier.getAsLong() == current.getPrimaryTerm()) {
+            return;
+        }
+        prepareAndUpload(primaryTermSupplier.getAsLong(), null);
+    }
+
+    private boolean prepareAndUpload(Long primaryTerm, Long generation) throws IOException {
+        try (Releasable ignored = writeLock.acquire()) {
+            if (generation == null || generation == current.getGeneration()) {
+                try {
+                    final TranslogReader reader = current.closeIntoReader();
+                    readers.add(reader);
+                    copyCheckpointTo(location.resolve(getCommitCheckpointFileName(current.getGeneration())));
+                    if (closed.get() == false) {
+                        logger.trace("Creating new writer for gen: [{}]", current.getGeneration() + 1);
+                        current = createWriter(current.getGeneration() + 1);
+                    }
+                } catch (final Exception e) {
+                    tragedy.setTragicException(e);
+                    closeOnTragicEvent(e);
+                    throw e;
+                }
+            } else if (generation < current.getGeneration()) {
+                return false;
+            }
+
+            // Do we need remote writes in sync fashion ?
+            // If we don't , we should swallow FileAlreadyExistsException while writing to remote store
+            // and also verify for same during primary-primary relocation
+            // Writing remote in sync fashion doesn't hurt as global ckp update
+            // is not updated in remote translog except in primary to primary recovery.
+            if (generation == null) {
+                if (closed.get() == false) {
+                    return upload(primaryTerm, current.getGeneration() - 1);
+                } else {
+                    return upload(primaryTerm, current.getGeneration());
+                }
+            } else {
+                return upload(primaryTerm, generation);
+            }
+        }
+    }
+
+    private boolean upload(Long primaryTerm, Long generation) throws IOException {
```

Comment:
Maybe check if lock is held by current thread. Please also add multi-threaded tests to ensure that replication requests and flush request don't get into concurrency issues